File: leetcode/easy/746-min-cost/dp.py
"""
  @author Vortexx2
  Problem 746 - Minimum Cost in Climbing Stairs

  Runtime - 93 ms
  Memory Usage - 14.4 MB
"""
from typing import List


class Solution:
  def minCostClimbingStairs(self, cost: List[int]) -> int:

    if len(cost) <= 1:
      return 0

    self.dp = [0, 0]
    return self.minCost(cost, len(cost))

  def minCost(self, cost: List[int], current: int) -> int:

    if current < len(self.dp):
      return self.dp[current]

    self.dp.append(min(self.minCost(cost, current - 2) +
                   cost[current - 2], self.minCost(cost, current - 1) + cost[current - 1]))

    return self.dp[current]

File: leetcode/easy/746-min-cost/dp.cpp
/**
 * @file dp.cpp
 * @author Vortexx2
 * @brief Problem 746 - Minimum Cost Climbing Stairs
 * @date 27-06-2022
 *
 * Runtime - 6 ms
 * Memory Usage - 14.1 MB
 */
#include <iostream>
#include <vector>

using namespace std;

class Solution {
 public:
  vector<int> dp = {0, 0};
  int minCostClimbingStairs(vector<int>& cost) {
    if (cost.size() <= 1) return 0;

    return minCost(cost, cost.size());
  }

  int minCost(vector<int>& cost, int current) {
    if (current < dp.size()) return dp[current];

    dp.push_back(min(minCost(cost, current - 2) + cost[current - 2],
                     minCost(cost, current - 1) + cost[current - 1]));

    return dp[current];
  }
};